features/timescale/src/main/java/org/opennms/netmgt/measurements/impl/TimescaleFetchStrategy.java

```diff
@@ -0,0 +1,452 @@
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Used to retrieve measurements from {@link org.opennms.newts.api.SampleRepository}.
+ *
+ * If a request to {@link #fetch} spans multiple resources, separate calls to
+ * the {@link SampleRepository} will be performed in parallel.
+ *
+ * Reading the samples and computing the aggregated values can be very CPU intensive.
+ * The "parallelism" attribute is used to set an upper limit on how may concurrent threads
+ * can be used to perform these calculations. By default, this is set to the number of
+ * cores, but can be reduced if the operator wishes to ensure cores are available
+ * for other purposes.
+ *
+ * @author jwhite
+ */
+public class TimescaleFetchStrategy implements MeasurementFetchStrategy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TimescaleFetchStrategy.class);
+
+    public static final long MIN_STEP_MS = SystemProperties.getLong("org.opennms.newts.query.minimum_step", 5L * 60L * 1000L);
+
+    public static final int INTERVAL_DIVIDER = SystemProperties.getInteger("org.opennms.newts.query.interval_divider", 2);
+
+    public static final long DEFAULT_HEARTBEAT_MS = SystemProperties.getLong("org.opennms.newts.query.heartbeat", 450L * 1000L);
+
+    public static final int PARALLELISM = SystemProperties.getInteger("org.opennms.newts.query.parallelism", Runtime.getRuntime().availableProcessors());
+
+    @Autowired
+    private Context m_context;
+
+    @Autowired
+    private ResourceDao m_resourceDao;
+
+    @Autowired
+    private SampleRepository m_sampleRepository;
+
+    private final ThreadFactory namedThreadFactory = new ThreadFactoryBuilder().setNameFormat("NewtsFetchStrateg-%d").build();
+
+    private final ExecutorService threadPool = Executors.newCachedThreadPool(namedThreadFactory);
+
+    // Used to limit the number of threads that are performing aggregation calculations in parallel
+    private final Semaphore availableAggregationThreads = new Semaphore(PARALLELISM);
+
+    @Autowired
+    private DataSource dataSource;
+    private Connection connection;
+
+    @Override
+    public FetchResults fetch(long start, long end, long step, int maxrows, Long interval, Long heartbeat, List<Source> sources, boolean relaxed) {
+        final LateAggregationParams lag = getLagParams(step, interval, heartbeat);
+        final Optional<Timestamp> startTs = Optional.of(Timestamp.fromEpochMillis(start));
+        final Optional<Timestamp> endTs = Optional.of(Timestamp.fromEpochMillis(end));
+        final Map<String, Object> constants = Maps.newHashMap();
+
+        FetchResults result;
+
+        // TODO Patrick: do db stuff proper.
+        try {
+
+            long computedStepInMillis = lag.getStep();
+            String computedStepInSeconds = (lag.getStep() / 1000 ) + " Seconds";
+            String resourceId = "response:127.0.0.1:response-time"; // TODO Patrick: deduct from sources
```

Comment:
@j-white we should discuss why the source.resourceId (eg. `node[localhost:1575287230289].responseTime[127.0.0.1]`) is different from the Newts resourceId (e.g. `response:127.0.0.1:response-time`). I'm not clear on this.

Reply:
Response time resources aren't actually associated with a node when they are stored i.e. (`response:127.0.0.1:response-time`), but we allow them to be retrieved by node for convenience.